File: 프로그래머스/1/42862. 체육복/체육복.java
import java.util.*;

class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        Arrays.sort(reserve);
        Arrays.sort(lost);
        Set<Integer> need = new HashSet<>();
        Set<Integer> extra = new HashSet<>();
        for(int i : reserve) extra.add(i);
        for(int i : lost) {
            if(extra.contains(i)) extra.remove(i); // 여별옷 가져와놓고 도둑맞은 사람
            else need.add(i);
        }
        int answer = n - need.size();

        for(int i : new TreeSet<>(extra)) {
            if(need.contains(i-1)){
                answer++;
                need.remove(i-1);
            } else if(need.contains(i+1)) {
                answer++;
                need.remove(i+1);
            }
        }
        return answer;
    }
}

File: 프로그래머스/1/42862. 체육복/체육복.cpp
#include <set>
#include <unordered_set>
#include <vector>
#include <algorithm> 
using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    // sort(lost.begin(), lost.end());
    // sort(reserve.begin(), reserve.end());
    unordered_set<int> need;
    set<int> extra;
    for (const int& i : reserve) extra.insert(i);
    for (const int& i : lost)
    {
        if (extra.find(i) != extra.end()) extra.erase(i);
        else need.insert(i);
    }
    int answer = n - static_cast<int>(need.size());
    for (const int& i : extra)
    {
        if (need.find(i-1) != need.end())
        {
            answer++;
            need.erase(i-1);
        } else if (need.find(i+1) != need.end())
        {
            answer++;
            need.erase(i+1);
        }
    }

    return answer;
}